File: 01/01.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Move struct {
	Direction string
	Steps     int
}

func parseMoves(lines []string) []Move {
	var moves []Move
	for _, line := range lines {
		dir := line[0:1]
		steps, err := strconv.Atoi(line[1:])
		if err != nil {
			continue
		}
		moves = append(moves, Move{Direction: dir, Steps: steps})
	}
	return moves
}

func part1(lines []string) int {
	moves := parseMoves(lines)

	position := 50
	zeros := 0
	for _, move := range moves {
		switch move.Direction {
		case "R":
			position += move.Steps
		case "L":
			position -= move.Steps
		}
		position %= 100
		if position == 0 {
			zeros++
		}
	}
	return zeros
}

func part2(lines []string) int {
	moves := parseMoves(lines)

	position := 50
	zeros := 0

	for _, move := range moves {
		step := 0
		switch move.Direction {
		case "R":
			step = 1
		case "L":
			step = -1
		}

		for i := 0; i < move.Steps; i++ {
			position += step
			position %= 100
			if position == 0 {
				zeros++
			}
		}
	}

	return zeros
}

func main() {
	data, err := os.ReadFile("01/input.txt")
	if err != nil {
		panic(err)
	}
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	result1 := part1(lines)
	fmt.Println("Part 1:", result1)

	result2 := part2(lines)
	fmt.Println("Part 1:", result2)

}

File: 01/01.py
#! /usr/bin/env python3


def part1(input_data):
    position = 50

    zeros = 0

    for line in input_data.splitlines():
        direction = line[0]
        value = line[1:]
        value = int(value)

        if direction == "R":
            position += value
        elif direction == "L":
            position -= value

        position %= 100

        if position == 0:
            zeros += 1

    return zeros


def part2(input_data):
    position = 50

    zeros = 0

    for line in input_data.splitlines():
        direction = line[0]
        value = line[1:]
        value = int(value)

        step = 0
        if direction == "R":
            step = 1
        elif direction == "L":
            step = -1

        while value > 0:
            position += step
            position %= 100
            if position == 0:
                zeros += 1

            value -= 1

    return zeros


if __name__ == "__main__":
    with open("01/sample.txt") as f:
        input_data = f.read()
        p1_zeros = part1(input_data)
        print(f"Part1 sample: {p1_zeros}")
        assert p1_zeros == 3

        p2_zeros = part2(input_data)
        print(f"Part2 sample: {p2_zeros}")
        assert p2_zeros == 6

    with open("01/input.txt") as f:
        input_data = f.read()
        p1_zeros = part1(input_data)
        print(f"Part1 input: {p1_zeros}")
        assert p1_zeros == 1100

        p2_zeros = part2(input_data)
        print(f"Part2 input: {p2_zeros}")
        assert p2_zeros == 6358
